### ISSUE TITLE
User friendly API for using pipelines on TimeSeries objects

### ISSUE THREAD

**Issue by pjm17971** [esnet/pond#22]
Here is an example of trying to use the current API (as it's exposed on TimeSeries.collapse())

The use case here is to take a TimeSeries, which is to be stacked, and find the appropiate scale. This has three parts to it:

 * For a subset of columns (those which will be charted in the up direction) we element by element collapse it to a single column which is the sum of the columns being collapse. i.e. up = ["oscars", "lhcone", "other"]. The output will be a single column ("value") which is the sum of oscars, lhcone and other, for each time.
 * Then for each of these, find the max down the column. This gives us our chart scale in the up direction. We repeat 1. and 2. for the down direction.
 * Since our chart will be scaled the same in both the up and down direction, find the max of the two max values. This becomes our scale.

To do this in code, within a React component, looks like this:

Firstly, we are going to keep the scale as state on the component. When the TimeSeries changes we recalculate the scale. This is so that we don't have to calculate the scale everytime the thing renders:

```
componentWillReceiveProps(nextProps) {
    const { series, up, down } = nextProps;
    if (!TimeSeries.is(this.props.series, series)) {
        this.updateChartScale(series, up, down);
    }
}
```

updateChartScale() does the actual code to get the new state:

```
updateChartScale(timeseries, up, down) {
    // A function to collapse columns to their sum, then find the max
    // value across the series
    const collapser = (columns, cb) => {
        timeseries.collapse(columns, "value", sum, false, sums =>
            cb(null, sums.max())
        );
    }

    // Use async to collect together the two callback results: the max
    // in the up and down direction, then do a standard React setState
    // to the max of those

    async.series([
        cb => collapser(up, cb),
        cb => collapser(down, cb)
    ], (err, results) => {
         this.setState({max: _.max(results)});
    });
}
```
Setting the state causes a re-render, which can then just use this.state.max in the YAxis code.

Pretty much a pure reactive approach. Sure.

Too hard? Yeah, probably.

An alternative I've been looking into is to have batch pipelines be optionally sync. Well to be clear they currently aren't async now, strictly speaking, but their API allows them to return results as they are built, and those results, in the future, might not happen in the same tick as the output call (`to()`).

The proposal is to add two additional functions to Pipeline, to complement `to()`. `toEventList()` and `toCollectionList()`. If you evoke these with a streaming source, it will throw. If you evoke them with a bounded source (e.g. a `TimeSeries`), it will run all of the source through the pipeline, then send a flush signal. When each `Event` (or `Collection`) is received by the `Out`, and that `Out` doesn't have an observer set (which will be the case when using the two new output methods), it will instead add the event (or collection) to a results accumulator on the `Pipeline` itself. When the flush is received by the `Out`, it will tell the `Pipeline` that the results are complete.

To the user this would look like this:
```
const timeseries = new TimeSeries(data);
const events = Pipeline()
    .from(timeseries)
    .emitOn("flush")
    .collapse(["in", "out"], "total", sum)
    .aggregate({total: max})
    .toEventList();
console.log("RESULT:", events);
```
Or for collections:
```
const timeseries = new TimeSeries(data);
const collections = Pipeline()
    .from(timeseries)
    .emitOn("flush")
    .groupBy(e => e.value() > 65 ? "high" : "low")
    .take(10)
    .toCollectionList();
console.log(collections);  // 2 collections (high and low)
```

----

The tricky thing here is what if this becomes async in the future.

This could happen in two ways:
 * The from() source could be async. For instance it could be a source requested off the network or from a file. If this was the case the `toEventList()` would return with an empty result set. Sometime later the source would have data and it would start to feed that data into the pipeline.
 * A part of the pipeline is async because it farms out work to workers. The first time this happens the stack will unwind and the pipeline will return, again with no results.

To deal with these I think you'd have to say that if you use the two new methods to return data immediately, the pipeline becomes sync (e.g. a flag is passed down the pipeline, async=false, and everything should honor that). I don't know of anyway to make a network call and actually block waiting for the response (and that would be a bad idea in a browser anyway), so likely such a source, if present, would throw based on the async flag. For file loading on node.js it could block with a *sync method, but again, not such a good idea. For farming out work, it could just not do that. No parallel for you.

----

On balance the pipeline async API is the right one for a) a reactive approach, b) streaming and c) async sources that aren't a TimeSeries you already have and d) future parallel operations. But, the real point here is if all you have a TimeSeries, and you just want to calculate something on it, it should be simpler.

Here's the first example again, assuming an implementation of collapse that uses the above changes:
```
updateChartScale(timeseries, up, down) {
    const maxUp = timeseries.collapse(up, "value", sum);
    const maxDown = timeseries.collapse(down, "value", sum);
    this.setState({max: Math.max(maxUp, maxDown)});
}
```

**Comment by jdugan1024** [comment]
+1 This seems like a reasonable compromise.  

With `async.series()` it isn't too bad to understand but quite a bit harder than the final example.  I guess the ultimate question for me is how much additional complexity does this add to the code overall.  It doesn't seem like too much but just checking.

**Comment by pjm17971** [comment]
This was included in v0.6.
